### DISCUSSION TITLE
Pre-filled combobox prevents seeing other items in menu without clearing field

### DISCUSSION THREAD

**Discussion by rosemaryluo** [twilio-labs/paste#2985] [Q&A] [answered]
Hi, it's me again with the same combobox, different issue.
We have a country selector that we enabled search functionality for since it's a pain to scroll through countries
However, we pre-populate this with a detected country e.g. United States
so when someone goes to try to change it, the default behavior of the dropdown is to filter by the input "United States" and not show all the other countries until the input is cleared. This is leading to a <50% success rate at actually selecting a different country.
We started chatting about workarounds, but first I want to confirm whether this is the intended behavior for this component.
Thank you!
<img width="396" alt="image" src="https://user-images.githubusercontent.com/87049408/214722449-2371ee60-505c-453c-95a0-e434f2574385.png">
<img width="660" alt="image" src="https://user-images.githubusercontent.com/87049408/214722867-cabe1889-72cd-4083-9945-8b1d5185e3c5.png">
<img width="651" alt="image" src="https://user-images.githubusercontent.com/87049408/214722933-95aa5bd1-d755-4c19-8781-c999c288d5cf.png">
<img width="632" alt="image" src="https://user-images.githubusercontent.com/87049408/214723027-612cf0e0-9072-4637-b031-0e9d2d778071.png">

**Comment by rosemaryluo** [comment]
Quick prototype of an alternative we are interested in exploring, so that we can get the countries to show up before deleting the pre-selected input: 
https://www.figma.com/proto/rok8UhaMG22Xqck3yanEHX/Verification-and-Ahoy-migration-to-Paste-(2022)?page-id=1173%3A104836&node-id=1241%3A106695&viewport=27%2C47%2C0.37&scaling=min-zoom&starting-point-node-id=1241%3A106695

**Comment by gloriliale** [comment, marked as answer]
Hi @rosemaryluo! Thanks for the question, I can definitely see this being difficult for a user. Let me consult the team and get back to you with our thoughts!

**Reply by rosemaryluo** [reply]
Thank you! Happy to join in with any context and provide any data, I tried to find an office hour slot but the next open slot I can attend will be 16th. We have this in a couple of spots and want to avoid doing a one-off alteration.
The Figma component only has the dropdown version and not the search version (without drop arrow, etc.), so I initially only noticed this because the live form was missing the dropdown arrow icon. I had assumed the search variety would look the same & wonder if this affects how other UXers are making assumptions about searchable combobox behavior.

**Reply by gloriliale** [reply]
Okay so I took another look, and I wasn't able to reproduce the issue. Could your team take a look at [this codesandbox](https://codesandbox.io/s/hopeful-cookies-h2qtkw?file=/src/index.tsx) to see if they're implementing it the same way?

**Reply by rosemaryluo** [reply]
From your sandbox, I could see the initial list, but after making a different selection, the same behavior happens.
https://user-images.githubusercontent.com/87049408/215229303-5e11e166-aa5a-4492-b31d-6e0bb9fbe054.mov

**Reply by gloriliale** [reply]
ah! thank you for the video, super helpful. I can see that it shows all the choices initially when it's pre-populated, but not after a choice has been made after. Thanks for helping make sure we were looking at the right issue! Consulting the team now.

**Reply by gloriliale** [reply]
Okay ONE more thing:

It looks like the behavior you're wanting, of having the whole item list show up when first clicking on the pre-populated field, is happening. After a different selection is made, then the list only shows that selected item until the field is cleared. Did you find that you need to show all the options after a user makes a different selection?

**Reply by rosemaryluo** [reply]
Having it show the first time is very helpful but it would be preferred to have consistent behavior when clicking in and wanting to make a change. My assumption would be that if they choose to click in again that they need to make another change.

**Reply by uditadhikari117** [reply]
Hello,

Just chiming in. I created a mockup with three permutations of combobox in this deploy preview link :
https://deploy-preview-3192.console.stage.twilio.com/account/ahoy
These permutations are : 
1.  Billing country with filtering logic and initial value
2. Billing country without filtering logic and initial value
3. Billing country without filtering logic and no initial value

Let me know Rosemary if what you would like to see isn't covered by one of these cases and then I can look further into this.
Gloriliale my PR is this : https://github.com/twilio-internal/console/pull/3192

**Reply by rosemaryluo** [reply]
@uditadhikari117 I'm looking for similar to #1 but not exactly the behavior you are showing. I would like the full list to show every time the user re-opens the dropdown after making a selection.

Looking to be able to do the following:
- click on dropdown (with pre-fill) - show full list of countries
- click on selection - dropdown closes
- click on dropdown (with new selection) - show full list of countries
- start typing in a new country - show filtered list of countries
- click on a selection - dropdown closes
- click on dropdown (with new selection) - show full list of countries
- (repeatable)

Attempted to recreate this flow in prototype:
https://user-images.githubusercontent.com/87049408/215556848-f5fa3c45-9577-48a4-8253-dfa0057c71be.mov

**Reply by uditadhikari117** [reply]
Ok after enough hacking I think I have what Rosemary wants. Rosemary please confirm. 
https://deploy-preview-3192.console.stage.twilio.com/account/ahoy

[gloriliale](https://github.com/gloriliale) what complicated this is anytime 'onSelectedItemChange' would run then 'onInputValueChange' would run right after. So, I am making sure that 'onInputValueChange' does not affect filtering when it receives an input that matches exactly one of the values in the list. 
The net effect is 'onSelectedItemChange' resets everything to a full list and 'onInputValueChange' then does no additional filtering. I think this is slightly hacky but open to your thoughts on this. However, I refrained from changing anything inside combobox.
PR -> https://github.com/twilio-internal/console/pull/3192/

**Reply by gloriliale** [reply]
Hi @uditadhikari117! That looks like it works great! I'd say it works for 99% of use cases.

The only catch is that if someone tries to select something that is already selected, which I would estimate is a true edge case (but let me know if I'm mistaken @rosemaryluo!), then `onSelectedItemChange` won't be triggered so they'll need to clear the field to see the full country list again. 

I truly think what you have will work wonderfully! If you want to check out another solution, where we address the rare edge case above, check out [this codesandbox](https://codesandbox.io/s/combobox-with-usecombobox-and-state-reducer-6pihbe) that makes use of the [useCombobox hook](https://paste.twilio.design/components/combobox#usecombobox-state-hook) that's also included in the Combobox package. Keep in mind that we warn consumers to **proceed with caution** using this hook, since it can be easy to break and difficult to debug. 

Thanks for reaching out!

**Reply by uditadhikari117** [reply]
Thank you @gloriliale. I think your solution is great as it removes that edge case. I think the edge case is hard to run into but when one does, it makes the usage of Combo Box a bit confusing. I will push the team towards your solution of using the `useCombobox` hook. I do have some maintenance concerns as our team is very backend heavy. However, I think it would be good experience for the team to use hooks.

**Reply by uditadhikari117** [reply]
I do want to spend some time understanding how the useCombobox works. If I have to guess it is using some sort of useReducer hook underneath. In the console zen repository I can only find the interface that the useCombobox hook subscribes to. Could I request the source code of useCombobox hook?
